### DISCUSSION TITLE
capnp transcoder and support in mcap cat --json output

### DISCUSSION THREAD

**Discussion by jon-chuang** [foxglove/mcap#1098] [General]
**Description**
We want to be able to decode capnp with dynamic schema in  `mcap-support` e.g. as is done for protobuf: https://github.com/foxglove/studio/blob/3767475a507097b3e4e046fc4d8b3a20135101f8/packages/mcap-support/src/parseChannel.ts#L125-L134

As a POC, we should first implement this in python and go for the CLI. The go API does not support reflection. I am thinking of a C-ABI FFI that we can bind to cgo. WASM compilation for browser can come later...

@jtbandes RE Slack lounge discussion

**Comment by foxhubber** [comment]
Linear: [FG-2433](https://linear.app/foxglove/issue/FG-2433/capnp-transcoder-and-support-in-mcap-cat-json-output-[foxglovemcap)

**Comment by jon-chuang** [comment]
Looking at all the possible pathways, while it may be fun to explore WASM compilation (see e.g. https://medium.com/cyberbotics/porting-a-c-rendering-engine-to-webassembly-9c32d76c31f1) or using cgo to bind to the JSON transcoder (https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/capnp/compat/json.h), given the poor support of Capnproto in various languages, especially for browser, It does seem to make much more sense to encode the original capnp recording into a compatible format like JSON (or CBOR for compactness) at time of storage. Anything that decodes capnproto in a browser environment will not be zero-copy anyway.

**Comment by defunctzombie** [comment]
@jon-chuang Please me mindful when using the bug label. The bug label is for broken functionality. If you believe this issue is a bug please explain what existing functionality that we've documented is not working, provide steps to reproduce, and describe what you expect to happen.

**Comment by jtbandes** [comment]
I'm going to convert this to a discussion since we haven't yet settled on the spec updates to "well-known" encodings for capnp usage.

**Comment by jon-chuang** [comment]
I have gotten it working on WASM, actually. Here is some sample code:

```js
    const buffer_test = new Uint8Array([1, 0, 0, 0, 2, 0, 0, 0]);
    console.log(Module.test(buffer_test))

    var enc = new TextEncoder();

    sampleMsgs = enc.encode(Module.buildSampleMessages());
    schemas = Module.buildSampleSchemas();

    console.log(sampleMsgs);
    console.log(schemas);

    var transcoder = new Module.SchemaTranscoder(schemas);

    var jsons = transcoder.encodeMessagesToJson(sampleMsgs, 0);
    console.log(jsons.get(0));
    console.log(jsons.get(1));
    console.log(jsons.get(2));
 ```
 
 with console output:
 
 
![image](https://user-images.githubusercontent.com/9093549/225010027-a31b624d-ed0b-41f4-88bb-e5fd6767ce44.png)

<details>
It currently does not work on nested schemas, unfortunately, resulting in the `42` value for the "value" key being blotted out with `{}`

I think this is JSON-specific. I am thinking of transcoding to a more compact format and then utilizing the existing tooling to translate to e.g. foxglove types.
</details>

**Comment by jon-chuang** [comment]
> haven't yet settled on the spec updates to "well-known" encodings for capnp usage

Could you elaborate on what this means? 

In particular:  spec updates, well-known.

I believe that we can just rely on the latest versions for the encoding version? I don't think capnp or msgpack have changed for a very long time.

**Reply by defunctzombie** [reply]
Similar to mime-types, image formats, video formats, we have an appendix to the mcap specification which reserves specific encoding values that correspond to well-known encodings. Found here: https://mcap.dev/specification/appendix.html and source here: https://github.com/foxglove/mcap/blob/main/docs/specification/appendix.md. Once we reserve a specific name for an encoding and schema, other tools can know that if they want to support users of that encoding this is what they should support.

Feel free to propose a PR to the mcap appendix with what language we should add to support capnp and msgpack. Both will need schema and message encoding consideration.

**Reply by jon-chuang** [reply]
Cool. Yes, I understand, it needs a name for schema, and schema data format, or just a link to the encoding for schemaless.
